clients/src/main/java/org/apache/kafka/clients/consumer/internals/MembershipManagerImpl.java

```diff
@@ -181,33 +307,469 @@ public void updateState(ConsumerGroupHeartbeatResponseData response) {
-        transitionTo(MemberState.FAILED);
+    public void transitionToFatal() {
+        log.error("Member {} transitioned to {} state", memberId, MemberState.FATAL);
+
+        // Update epoch to indicate that the member is not in the group anymore, so that the
+        // onPartitionsLost is called to release assignment.
+        memberEpoch = ConsumerGroupHeartbeatRequest.LEAVE_GROUP_MEMBER_EPOCH;
+        invokeOnPartitionsRevokedOrLostToReleaseAssignment();
+
+        clearAssignedTopicNamesCache();
+        transitionTo(MemberState.FATAL);
     }
 
+    /**
+     * {@inheritDoc}
+     */
     @Override
-    public boolean shouldSendHeartbeat() {
-        return state() != MemberState.FAILED;
+    public void transitionToJoining() {
+        resetEpoch();
+        transitionTo(MemberState.JOINING);
+        clearAssignedTopicNamesCache();
     }
 
     /**
-     * Transition to {@link MemberState#STABLE} only if there are no target assignments left to
-     * reconcile. Transition to {@link MemberState#RECONCILING} otherwise.
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<Void> leaveGroup() {
+        transitionTo(MemberState.PREPARE_LEAVING);
+
+        CompletableFuture<Void> callbackResult = invokeOnPartitionsRevokedOrLostToReleaseAssignment();
+        callbackResult.whenComplete((result, error) -> {
+
+            // Clear the subscription, no matter if the callback execution failed or succeeded.
+            subscriptions.assignFromSubscribed(Collections.emptySet());
+
+            // Transition to ensure that a heartbeat request is sent out to effectively leave the
+            // group (even in the case where the member had no assignment to release or when the
+            // callback execution failed.)
+            transitionToSendingLeaveGroup();
+
+        });
+
+        clearAssignedTopicNamesCache();
+
+        // Return callback future to indicate that the leave group is done when the callbacks
+        // complete, without waiting for the heartbeat to be sent out. (Best effort to send it
+        // but do not hold the leave group operation for it)
+        return callbackResult;
+    }
+
+    /**
+     * Release member assignment by calling the user defined callbacks for onPartitionsRevoked or
+     * onPartitionsLost.
+     * <ul>
+     *     <li>If the member is part of the group (epoch > 0), this will invoke onPartitionsRevoked.
+     *     This will be the case when releasing assignment because the member is intentionally
+     *     leaving the group (after a call to unsubscribe)</li>
+     *
+     *     <li>If the member is not part of the group (epoch <=0), this will invoke onPartitionsLost.
+     *     This will be the case when releasing assignment after being fenced .</li>
+     * </ul>
+     *
+     * @return Future that will complete when the callback execution completes.
+     */
+    private CompletableFuture<Void> invokeOnPartitionsRevokedOrLostToReleaseAssignment() {
```

Comment:
is there any reason we want to combine onPartitionsLost and onPartitionsRevoke into a single function?  Couldn't we directly invoke `invokeOnPartitionsRevoke` or `invokeOnPartitionsLost`? i.e. by the time the consumer is fenced or would know we need to invoke onPartitionsLost

Reply:
The reason is the leave group logic. On leave group it could be lost or revoked, depending on the epoch. On fence or fatal is always lost. I was just reusing the same func for convenience, but will change the fence and fatal transitions to directly invoke the `onPartitionsLost` just to make the intention clearer.